### ISSUE TITLE
Disable Token system

### ISSUE THREAD

**Issue by PhilipVinc** [PhilipVinc/mpi4jax#17]
When using only operations like all reduce, which have just one result which depends on the others, you don't really need the token system.

As such, it would be nice if it could be possible to disable it, maybe with `token=false`, so that operations do not return a token. 
@dionhaefner  What do you think?

**Comment by dionhaefner** [comment]
Problems with this:

- You still need it if you do more than one operation in a row:

   ```python
   @jax.jit
   def myfunc(arr1, arr2):
       # don't do this!
       res1 = Allreduce(arr1, MPI.SUM)
       res2 = Allreduce(arr2, MPI.SUM)
       return res1, res2
   ```

   XLA is free to re-order the two `Allreduce` calls on some but not all processes which causes garbage results.

- We would need two versions of each primitive implementation (one that returns a token and one that doesn't), essentially doubling the code.

I don't really see the harm in writing `res, = Allreduce(...)` when you don't care about tokens. But if you really want this I would propose the following API:

```python
# no token in -> no token out
res = Allreduce(arr)

# token in  -> token out
token = create_token()
res, token = Allreduce(arr, token=token)
res, token = Allreduce(arr, token=token)
```

**Comment by PhilipVinc** [comment]
indeed you are right.
But then I need to find a way to batch and to compute gradients of allreduce. I had completely forgotten that it is essential to what I need mpi4jax for...

**Comment by dionhaefner** [comment]
You can try to remove these lines:

https://github.com/PhilipVinc/mpi4jax/blob/9e43d39c80bc95218967c04bcb61120d445947a2/mpi4jax/collective_ops/allreduce.py#L48-L49

and require that the token is created externally and passed in from the outside. Then you just need to tell the batching somehow that you only want to batch along the first input and output (and analogously with gradients). I don't know how / if that works, I haven't used that part of JAX yet.

**Comment by PhilipVinc** [comment]
I tried to ask google people about actually defining the adjoint for `create_token`. See [google/jax#4292](https://github.com/google/jax/issues/4292)

**Comment by PhilipVinc** [comment]
That would solve the gradient, which I can make do for now.
Batching is indeed a more complex problem. We would need an array of tokens, but jax does not support that ATM

**Comment by dionhaefner** [comment]
If you look at how `create_token` is defined in JAX, it's just a very thin wrapper primitive:

https://cs.opensource.google/jax/jax/+/master:jax/lax/lax.py;l=5710?q=create_token&sq=&ss=jax%2Fjax

So you could define your own version of `create_token` that has a gradient defined:

```python
def my_create_token(x):
  return my_create_token_p.bind(x)

my_create_token_p = Primitive("create_token_with_grad")
my_create_token_p.def_impl(partial(xla.apply_primitive, my_create_token_p))
my_create_token_p.def_abstract_eval(lambda _: abstract_token)
xla.translations[my_create_token_p] = lambda c, _: xops.CreateToken(c)

def my_gradient_rule(...):
    # do what you need to do

batching.primitive_batchers[my_create_token_p] = my_gradient_rule
```

**Comment by PhilipVinc** [comment]
Good idea.
So I avoid type pirating

**Comment by dionhaefner** [comment]
I like the suggestion to use the `argnums` or maybe `has_aux` arguments to `grad`. Did you try that?

**Comment by PhilipVinc** [comment]
Yep, see my reply in the jax issue.
It does not work because our MPI operations introduce a relationship between tokens and input/output of mpi operations (at least, in the eyes of the tracer)

**Comment by dionhaefner** [comment]
OK, but it works if you pass in the token from the outside, right? So it could be an option to create a token beforehand if you want to use grad?

**Comment by PhilipVinc** [comment]
yes, it does.
The problem of doing that is that it complicates a lot some code I have.
I really dislike doing that.

**Comment by dionhaefner** [comment]
Alright. I guess we can keep the current workaround for now and see if it creates any problems down the line.
